File: src/server/Model/Compiler/SubPanel.cs
// Copyright (c) 2020, UW Medicine Research IT, University of Washington
// Developed by Nic Dobbins and Cliff Spital, CRIO Sean Mooney
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
using System;
using System.Collections.Generic;
using System.Linq;

namespace Model.Compiler
{
    public class SubPanel : BaseSubPanel
    {
        public IEnumerable<PanelItem> PanelItems { get; set; }

        public bool HasCountFilter
        {
            get
            {
                return MinimumCount > 1;
            }
        }

        public SubPanel()
        {
            PanelItems = new List<PanelItem>();
        }
    }
}

File: src/ui-client/src/models/panel/SubPanel.ts
/* Copyright (c) 2021, UW Medicine Research IT, University of Washington
 * Developed by Nic Dobbins and Cliff Spital, CRIO Sean Mooney
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */ 

import { DateFilter, DateIncrementType } from './Date';
import { PanelItem, PanelItemDTO, panelItemToDto } from './PanelItem'

export interface BaseSubPanel {
    dateFilter: DateFilter;
    id: string;
    index: number;
    includeSubPanel: boolean;
    minimumCount: number;
    panelIndex: number;
    joinSequence: SubPanelJoinSequence;
}

export interface SubPanelDTO extends BaseSubPanel {
    panelItems: PanelItemDTO[];
 }

export interface SubPanel extends BaseSubPanel {
    panelItems: PanelItem[];
    joinSequenceEventType?: JoinSequenceEventType;
}

export interface JoinSequenceEventType {
    id: number;
    name: string;
}

export const subPanelToDto = (subpanel: SubPanel): SubPanelDTO => {
    return {
        dateFilter: subpanel.dateFilter,
        id: subpanel.id,
        includeSubPanel: subpanel.includeSubPanel,
        index: subpanel.index,
        joinSequence: subpanel.joinSequence,
        minimumCount: subpanel.minimumCount,
        panelIndex: subpanel.panelIndex,
        panelItems: subpanel.panelItems.map(i => panelItemToDto(i))
    }
}

export interface SubPanelJoinSequence {
    increment: number | null;
    dateIncrementType: DateIncrementType;
    sequenceType: SequenceType;
}

export enum SequenceType {
    Encounter = 0,
    PlusMinus = 1,
    WithinFollowing = 2,
    AnytimeFollowing = 3,
    Event = 4
}